pkg/registry/batch/job/strategy.go

```diff
@@ -331,7 +338,30 @@ func (jobStatusStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.
 }
 
 func (jobStatusStrategy) ValidateUpdate(ctx context.Context, obj, old runtime.Object) field.ErrorList {
-	return batchvalidation.ValidateJobUpdateStatus(obj.(*batch.Job), old.(*batch.Job))
+	newJob := obj.(*batch.Job)
+	oldJob := old.(*batch.Job)
+
+	opts := getStatusValidationOptions(newJob, oldJob)
+	return batchvalidation.ValidateJobUpdateStatus(newJob, oldJob, opts)
+}
+
+// getStatusValidationOptions returns validation options for Job status
+func getStatusValidationOptions(newJob, oldJob *batch.Job) batchvalidation.JobStatusValidationOptions {
+	if utilfeature.DefaultFeatureGate.Enabled(features.JobSuccessPolicy) {
+		isIndexed := ptr.Deref(newJob.Spec.CompletionMode, batch.NonIndexedCompletion) == batch.IndexedCompletion
+		oldJobHasSuccessCriteriaMet := batchvalidation.IsJobSuccessCriteriaMet(oldJob)
+		isJobSuccessCriteriaChanged := oldJobHasSuccessCriteriaMet != batchvalidation.IsJobSuccessCriteriaMet(newJob)
+		isJobFailedChanged := batchvalidation.IsJobFailed(oldJob) != batchvalidation.IsJobFailed(newJob)
+		isFailureTargetChanged := batchvalidation.IsJobFailureTarget(oldJob) != batchvalidation.IsJobFailureTarget(newJob)
+
+		return batchvalidation.JobStatusValidationOptions{
```

Comment:
#123273 is adding all these options because it's tightening existing validation.

In this case, you are adding validation for new fields, so you don't need options. You should always validate when the feature is enabled.

Reply:
@alculquicondor Actually, we decided to take the same approach as #123273 based on https://github.com/kubernetes/kubernetes/pull/123412#discussion_r1508250675.

If my understanding isn't correct, please let me know.

Reply:
It might be ok to have the options, but they should always be true when the feature is enabled.

Reply:
@alculquicondor Another discussion is here: https://github.com/kubernetes/kubernetes/pull/123412#discussion_r1511671300

It seems that it might be better to detect changes in the strategy. @alculquicondor @atiratree @deads2k WDYT?

Reply:
Detecting changes is only necessary when tightening validation, not for a new feature with new fields

Reply:
True, and we are only adding a new validations. Even though we depend on the `completionMode`, it cannot be mutated later. So we are not breaking existing jobs. And new ones should validate correctly unconditionally.

Reply:
I see. Let me re-implement these. Thanks for everyone.

Reply:
Done.